magicblock-processor/src/scheduler/mod.rs

```diff
@@ -0,0 +1,219 @@
+use log::info;
+use magicblock_core::link::transactions::{
+    ProcessableTransaction, TransactionToProcessRx,
+};
+use magicblock_ledger::LatestBlock;
+use solana_program_runtime::loaded_programs::ProgramCache;
+use state::TransactionSchedulerState;
+use tokio::{
+    runtime::Builder,
+    sync::mpsc::{channel, Receiver, Sender},
+};
+
+use crate::executor::{SimpleForkGraph, TransactionExecutor};
+
+/// The central transaction scheduler responsible for distributing work to a
+/// pool of `TransactionExecutor` workers.
+///
+/// This struct acts as the single entry point for all transactions entering the processing
+/// pipeline. It receives transactions from a global queue and dispatches them to available
+/// worker threads for execution or simulation.
+pub struct TransactionScheduler {
+    /// Manages the state of all executors, including locks and blocked transactions.
+    coordinator: ExecutionCoordinator,
+    /// The receiving end of the global queue for all new transactions.
+    transactions_rx: TransactionToProcessRx,
+    /// A channel that receives readiness notifications from workers,
+    /// indicating they are free to accept new work.
+    ready_rx: Receiver<ExecutorId>,
+    /// A list of sender channels, one for each `TransactionExecutor` worker.
+    executors: Vec<Sender<ProcessableTransaction>>,
+    /// A handle to the globally shared cache for loaded BPF programs.
+    program_cache: Arc<RwLock<ProgramCache<SimpleForkGraph>>>,
+    /// A handle to the globally shared state of the latest block.
+    latest_block: LatestBlock,
+    /// A shared atomic counter for ordering transactions within a single slot.
+    index: Arc<AtomicUsize>,
+}
+
+impl TransactionScheduler {
+    /// Creates and initializes a new `TransactionScheduler` and its associated pool of workers.
+    ///
+    /// This function performs the initial setup for the entire transaction processing pipeline:
+    /// 1.  Prepares the shared program cache and ensures necessary sysvars are in the `AccountsDb`.
+    /// 2.  Creates a pool of `TransactionExecutor` workers, each with its own dedicated channel.
+    /// 3.  Spawns each worker in its own OS thread for maximum isolation and performance.
+    pub fn new(executors: u32, state: TransactionSchedulerState) -> Self {
+        let count = executors.min(MAX_SVM_EXECUTORS) as usize;
+        let index = Arc::new(AtomicUsize::new(0));
+        let mut executors = Vec::with_capacity(count);
+
+        // Create the back-channel for workers to signal their readiness.
+        let (ready_tx, ready_rx) = channel(count);
+        // Perform one-time setup of the shared program cache and sysvars.
+        let program_cache = state.prepare_programs_cache();
+        state.prepare_sysvars();
+
+        for id in 0..count {
+            // Each executor has a channel capacity of 1, as it
+            // can only process one transaction at a time.
+            let (transactions_tx, transactions_rx) = channel(1);
```

Comment:
Could we extract this into separate const within function? It is highly important that no one ever sets anything rather than 1 here or the whole thing breaks.